pkcs5/src/pbes2/encryption.rs

```diff
@@ -48,6 +49,71 @@ fn cbc_decrypt<'a, C: BlockCipherDecrypt + BlockCipher + KeyInit>(
         .map_err(|_| Error::EncryptFailed)
 }
 
+fn gcm_encrypt<'a, C>(
+    es: EncryptionScheme,
+    key: EncryptionKey,
+    iv: &[u8],
+    buffer: &'a mut [u8],
+    pos: usize,
+) -> Result<&'a [u8]>
+where
+    C: BlockCipher + BlockSizeUser<BlockSize = U16> + GcmKeyInit + BlockCipherEncrypt,
+    aes_gcm::AesGcm<C, U12>: GcmKeyInit,
+{
+    const TAG_SIZE: usize = 16;
+    const NONCE_SIZE: usize = 12;
+
+    if buffer.len() < TAG_SIZE + pos {
+        return Err(Error::EncryptFailed);
+    }
+    let gcm = <aes_gcm::AesGcm<C, U12> as GcmKeyInit>::new_from_slice(key.as_slice())
+        .map_err(|_| es.to_alg_params_invalid())?;
+    let nonce: [u8; NONCE_SIZE] = iv.try_into().map_err(|_| es.to_alg_params_invalid())?;
+    let nonce = Nonce::from(nonce);
+    let tag = gcm
+        .encrypt_in_place_detached(&nonce, &[], &mut buffer[..pos])
+        .map_err(|_| Error::EncryptFailed)?;
+    buffer[pos..].copy_from_slice(tag.as_ref());
+    Ok(&buffer[0..pos + TAG_SIZE])
```

Comment:
```suggestion
fn gcm_encrypt<'a, C, NonceSize, TagSize>(
    es: EncryptionScheme,
    key: EncryptionKey,
    nonce: Nonce<C>,
    buffer: &'a mut [u8],
    pos: usize,
) -> Result<&'a [u8]>
where
    C: BlockCipher + BlockSizeUser<BlockSize = U16> + GcmKeyInit + BlockCipherEncrypt,
    aes_gcm::AesGcm<C, NonceSize, TagSize>: GcmKeyInit,
    TagSize: TagSize,
{
    const TAG_SIZE: usize = 16;
    const NONCE_SIZE: usize = 12;

    if buffer.len() < TagSize::USIZE + pos {
        return Err(Error::EncryptFailed);
    }
    let gcm = <aes_gcm::AesGcm<C, NonceSize, TagSize> as GcmKeyInit>::new_from_slice(key.as_slice())
        .map_err(|_| es.to_alg_params_invalid())?;
    let tag = gcm
        .encrypt_in_place_detached(&nonce, &[], &mut buffer[..pos])
        .map_err(|_| Error::EncryptFailed)?;
    buffer[pos..].copy_from_slice(tag.as_ref());
    Ok(&buffer[0..pos + TagSize::U8])
```

Reply:
That doesn't compile directly but I made effectively this change (along with similar for GCM decrypt) in 
603bbb4 PTAL

Reply:
I'm not surprised, I didn't run this through a compiler, and made the edits in GitHub's UI.